pkg/kube/krt/mergejoin.go

```diff
@@ -0,0 +1,315 @@
+	return j.syncer.WaitUntilSynced(s)
+}
+
+func (j *mergejoin[T]) List() []T {
+	j.mu.RLock()
+	defer j.mu.RUnlock()
+
+	return maps.Values(j.outputs)
+}
+
+func (j *mergejoin[T]) GetKey(k string) *T {
+	j.mu.RLock()
+	defer j.mu.RUnlock()
+
+	rf, f := j.outputs[Key[T](k)]
+	if f {
+		return &rf
+	}
+	return nil
+}
+
+func (j *mergejoin[T]) onSubCollectionEventHandler(o []Event[T]) {
+	var events []Event[T]
+	j.mu.Lock()
+	defer j.mu.Unlock()
+	// Between the events being enqueued and now, the input may have changed. Update with latest info.
+
+	items := j.refreshEventsLocked(o)
+
+	for _, ev := range items {
+		obj := ev.Latest()
+		objKey := getTypedKey(obj)
+		if ev.Event == controllers.EventDelete {
+			oldRes, f := j.outputs[objKey]
+			if !f {
+				j.log.WithLabels("iKey", objKey).Errorf("invalid event, deletion of non-existent object")
+				continue
+			}
+			e := Event[T]{
+				Event: controllers.EventDelete,
+				Old:   &oldRes,
+			}
+			events = append(events, e)
+			delete(j.outputs, objKey)
+			for _, index := range j.indexes {
+				index.delete(oldRes, objKey)
+			}
+			if j.log.DebugEnabled() {
+				j.log.WithLabels("res", objKey).Debugf("handled delete")
+			}
+		} else {
+			// We can trust these events as authoritative because we checked the state of the collections
+			// in refreshEvents.
+			newObj := ev.New
+			oldObj := ev.Old
+			if oldObj != nil && newObj != nil {
+				// Update event
+				if Equal(ev.New, ev.Old) {
+					// NOP change, skip
+					continue
+				}
+				j.outputs[objKey] = *newObj
+			} else if newObj != nil {
+				// Add event
+				j.outputs[objKey] = *newObj
+			} else {
+				if oldObj == nil && EnableAssertions {
+					panic(fmt.Sprintf("newObj and oldObj are both nil; how did we get here? for key %v", objKey))
+				}
+				j.log.WithLabels("iKey", objKey).Errorf("invalid event, Old has value but New is nil for non Delete event")
+				continue
+			}
+		}
+
+		for _, index := range j.indexes {
+			index.update(ev, objKey)
+		}
+
+		if j.log.DebugEnabled() {
+			j.log.WithLabels("res", objKey, "type", ev.Event).Debugf("handled")
+		}
+		events = append(events, ev)
+	}
+
+	// Short circuit if we have nothing to do
+	if len(events) == 0 {
+		return
+	}
+	if j.log.DebugEnabled() {
+		j.log.WithLabels("events", len(events)).Debugf("calling handlers")
+	}
+	j.eventHandlers.Distribute(events, !j.HasSynced())
+}
+
+func (j *mergejoin[T]) refreshEventsLocked(items []Event[T]) []Event[T] {
+	// Refreshing events is different for merged collections because we need to
+	// calculate the merge for the key of each item.
+	for idx, ev := range items {
+		iKey := getTypedKey(ev.Latest())
+		iObj := j.calculateMerged(string(iKey))
+		if iObj == nil {
+			ev.Event = controllers.EventDelete
+			if ev.Old == nil {
+				// This was an add, now its a Delete (because state changed between enqueue and now).
+				// Make sure we don't have Old and New nil, which we claim to be illegal.
+				ev.Old = ev.New
```

Comment:
ev.New could be outdated though right? Or do we just need to set ev.Old to be a non-nil value so we can get the key

Reply:
Yep exactly; we should reset ev.New to the value in our outputs later